qutip/solver/heom/bofin_baths.py

```diff
@@ -337,6 +346,181 @@ def combine(cls, exponents, rtol=1e-5, atol=1e-7):
+
+    def _bose_einstein(self, w):
+        """
+        Calculates the bose einstein distribution for the
+        temperature of the bath.
+
+        Parameters
+        ----------
+        w: float or array
+            Energy of the mode
+
+        Returns
+        -------
+        The population of the mode with energy w
+        """
+
+        if self.T is None:
+            raise NotImplementedError(
+                "Bath temperature must be specified for this operation")
+        if self.T == 0:
+            return np.zeros_like(w)
+
+        w = np.array(w, dtype=float)
+        with np.errstate(divide='ignore'):  # return inf if w=0
+            return (1 / (np.exp(w / self.T) - 1))
+
+    def power_spectrum(self, w):
+        """
+        Calculates the power spectrum from the spectral density using the
+        fluctuation dissipation theorem. This calculation requires that the
+        temperature of the bath (`self.T`) is specified.
+
+        Parameters
+        ----------
+        w: float or array
+            Energy of the mode
+
+        Returns
+        ----------
+        The power spectrum of the mode with energy w
+        """
+
+        # For w=0, the result would have the form 0 * inf
+        # We shift w by an epsilon to get the value in the limit w->0
+        w = np.array(w, dtype=float)
+        w[w == 0.0] += 1e-6
+
+        S = (2 * np.sign(w) * self.spectral_density(np.abs(w)) *
+             (self._bose_einstein(w) + 1))
+        return S
+
+    def correlation_function_approx(self, t):
+        """
+        Computes the correlation function from the exponents. This is the
+        approximation for the correlation function that is used in the HEOM
+        construction.
+
+        Parameters
+        ----------
+        t: float or array
+            time to compute correlations
+
+        Returns
+        ----------
+        The correlation function of the bath at time t
+        """
+
+        corr = np.zeros_like(t, dtype=complex)
+        for exp in self.exponents:
+            if (
+                exp.type == BathExponent.types['R'] or
+                exp.type == BathExponent.types['RI']
+            ):
+                corr += exp.ck * np.exp(-exp.vk * t)
+            if exp.type == BathExponent.types['I']:
+                corr += 1j*exp.ck * np.exp(-exp.vk * t)
+            if exp.type == BathExponent.types['RI']:
+                corr += 1j*exp.ck2 * np.exp(-exp.vk * t)
+        return corr
+
+    def power_spectrum_approx(self, w):
+        """
+        Calculates the power spectrum from the exponents
+        of the bosonic bath.
+
+        Parameters
+        ----------
+        w: float or array
+            Energy of the mode
+
+        Returns
+        ----------
+        The power spectrum of the mode with energy w
+        """
+
+        S = np.zeros_like(w, dtype=float)
+        for exp in self.exponents:
+            ck = exp.ck or 0
+            ck2 = exp.ck2 or 0
```

Comment:
Please check the bath type here rather than the values of `ck` and `ck2` like we do everywhere else.

Reply:
Sorry for the delay, This is to set the value of the ck from None to 0 to use it in calculation, we do check the bath.exponent below like everywhere else, but if we just check the bath exponent then some cks will be None. Could move this bit to the exponent types that have Nones for ck or ck2